### PR TITLE
ThumbRelocator: Implement handling of VLDR

### LINKED ISSUE #408: vldr (arm thumb) instruction translation error
My function header has "vldr" instruction. when i hook it, crashed! Function header as follow:

00ccd434 <xxxxxxx>:
ccd434: ed9f 1a1a   **vldr  s2, [pc, #104]**  
ccd438: ee00 1a10   vmov  s0, r1
ccd43c: ee20 2a01   vmul.f32  s4, s0, s2
ccd440: eebd 2ac2   vcvt.s32.f32  s4, s4

i checked coded, it don't translate vldr instruction,need fix

### PR DESCRIPTION
**Pull request by bigboysun**
fixed https://github.com/frida/frida-gum/issues/408

### CODE REVIEW

File: gum/arch-arm/gumthumbrelocator.c [hunk 1]
```diff
@@ -583,6 +588,30 @@ gum_thumb_relocator_rewrite_ldr (GumThumbRelocator * self,
...
+  if (src->type != ARM_OP_MEM || src->mem.base != ARM_REG_PC)
+    return FALSE;
+
+  absolute_pc = ctx->pc & ~((GumAddress) (4 - 1));
+  absolute_pc += src->mem.disp;
+
+  gum_thumb_writer_put_push_regs(ctx->output, 1, ARM_REG_R0);
+
+  gum_thumb_writer_put_ldr_reg_address (ctx->output, ARM_REG_R0, absolute_pc);
+  gum_thumb_writer_put_vldr_reg_reg_offset (ctx->output, dst->reg, ARM_REG_R0, 0);
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.c:608]: Exceeds 80 cols.

File: gum/arch-arm/gumthumbwriter.c [hunk 2]
```diff
@@ -788,6 +788,50 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
...
+{
+  GumArmRegInfo src_reg_info;
+  GumArmRegInfo dst_reg_info;
+
+  // cccc cccc cycc ssss dddd cccx oooo oooo
+  // c: opt code fixed
+  // y: dest reg index, bit 1 
+  // s: source reg index
+  // d: dest reg index, bit 2-5
+  // x: 0 f32 load  1 f64 load
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:805]: All comments should be C-style.

E.g.:

```C
  /*
   * Bla bla.
   * Some more text.
   */
```

File: gum/arch-arm/gumthumbwriter.c [hunk 3]
```diff
@@ -788,6 +788,50 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
...
+  uint32_t code = 0xed900a00;
+  gboolean is_float = TRUE;
+
+  gum_arm_reg_describe (src_reg, &src_reg_info);
+  gum_arm_reg_describe (dst_reg, &dst_reg_info);
+
+  code |= (src_offset >> 2) & 0xff;
+  code |= (src_reg_info.index << 16);
+
+  if (dst_reg_info.meta >= GUM_ARM_MREG_D0 && dst_reg_info.meta <= GUM_ARM_MREG_D31)
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:816]: Exceeds 80 cols.

File: tests/core/arch-arm/thumbrelocator.c [hunk 4]
```diff
@@ -84,6 +86,89 @@ TESTCASE (handle_extended_instructions)
   g_assert_cmpint (memcmp (fixture->output, input, sizeof (input)), ==, 0);
 }
 
+TESTCASE (vldrpc_t1_should_be_rewritten)
+{
+  const guint16 input[] = {
+    GUINT16_TO_LE (0xeddf), 
+    GUINT16_TO_LE (0x0a00), /*vldr  s1, [pc, #0] */
```
Comment (oleavr) [inline tests/core/arch-arm/thumbrelocator.c:93]:
```suggestion
    GUINT16_TO_LE (0x0a00), /* vldr  s1, [pc, #0] */
```

File: tests/core/arch-arm/thumbrelocator.c [hunk 5]
```diff
@@ -84,6 +86,89 @@ TESTCASE (handle_extended_instructions)
...
+  };
+  const guint16 expected_output_instructions[] = {
+    GUINT16_TO_LE (0xb401), /* push {r0}          */
+    GUINT16_TO_LE (0x4802), /* ldr  r0, [pc, #8] */
```
Comment (oleavr) [inline tests/core/arch-arm/thumbrelocator.c:97]:
```suggestion
    GUINT16_TO_LE (0x4802), /* ldr  r0, [pc, #8]  */
```

File: tests/core/arch-arm/thumbrelocator.c [hunk 6]
```diff
@@ -84,6 +86,89 @@ TESTCASE (handle_extended_instructions)
...
+
+  guint32 calculated_pc;
+  const cs_insn * insn = NULL;
+
+  SETUP_RELOCATOR_WITH (input);
+
+  memcpy (expected_output, expected_output_instructions,
+      sizeof (expected_output_instructions));
+  calculated_pc = (fixture->rl.input_pc + 4) & ~(4 - 1);
+#if G_BYTE_ORDER == G_LITTLE_ENDIAN
```
Comment (oleavr) [inline tests/core/arch-arm/thumbrelocator.c:115]: I suppose we don't need two codepaths here (and below), we can just assign the value without converting as we want to write the value using the native endian?

File: tests/core/arch-arm/thumbrelocator.c [hunk 7]
```diff
@@ -84,6 +86,89 @@ TESTCASE (handle_extended_instructions)
...
+  g_assert_cmpuint (gum_thumb_relocator_read_one (&fixture->rl, &insn), ==, 4);
+  g_assert_cmpint (insn->id, ==, ARM_INS_VLDR);
+  g_assert_true (gum_thumb_relocator_write_one (&fixture->rl));
+  gum_thumb_writer_flush (&fixture->tw);
+
+  g_assert_cmpint (memcmp (fixture->output, expected_output,
+      sizeof (expected_output)), ==, 0);
+}
+
+
```
Comment (oleavr) [inline tests/core/arch-arm/thumbrelocator.c:171]:
```suggestion
```

File: gum/arch-arm/gumarmreg.c [hunk 8]
```diff
@@ -14,25 +14,47 @@ gum_arm_reg_describe (arm_reg reg,
   {
     ri->meta = GUM_ARM_MREG_R0 + (reg - ARM_REG_R0);
     ri->width = 32;
+
```
Comment (oleavr) [inline gum/arch-arm/gumarmreg.c:17]:
```suggestion
```

File: gum/arch-arm/gumarmreg.c [hunk 9]
```diff
@@ -14,25 +14,47 @@ gum_arm_reg_describe (arm_reg reg,
...
+    ri->index = ri->meta - GUM_ARM_MREG_R0;
   }
   else if (reg == ARM_REG_SP)
   {
     ri->meta = GUM_ARM_MREG_SP;
     ri->width = 32;
+
```
Comment (oleavr) [inline gum/arch-arm/gumarmreg.c:24]:
```suggestion
```

File: gum/arch-arm/gumarmreg.c [hunk 10]
```diff
@@ -14,25 +14,47 @@ gum_arm_reg_describe (arm_reg reg,
...
+    ri->index = ri->meta - GUM_ARM_MREG_R0;
   }
   else if (reg == ARM_REG_LR)
   {
     ri->meta = GUM_ARM_MREG_LR;
     ri->width = 32;
+
```
Comment (oleavr) [inline gum/arch-arm/gumarmreg.c:31]:
```suggestion
```

File: gum/arch-arm/gumarmreg.c [hunk 11]
```diff
@@ -14,25 +14,47 @@ gum_arm_reg_describe (arm_reg reg,
...
+    ri->index = ri->meta - GUM_ARM_MREG_R0;
   }
   else if (reg == ARM_REG_PC)
   {
     ri->meta = GUM_ARM_MREG_PC;
     ri->width = 32;
+
```
Comment (oleavr) [inline gum/arch-arm/gumarmreg.c:38]:
```suggestion
```

File: gum/arch-arm/gumarmreg.c [hunk 12]
```diff
@@ -14,25 +14,47 @@ gum_arm_reg_describe (arm_reg reg,
...
+    ri->index = ri->meta - GUM_ARM_MREG_R0;
+  }
+  else if (reg >= ARM_REG_S0 && reg <= ARM_REG_S31)
+  {
+    ri->meta = GUM_ARM_MREG_S0 + (reg - ARM_REG_S0);
+    ri->width = 32;
+
```
Comment (oleavr) [inline gum/arch-arm/gumarmreg.c:45]:
```suggestion
```

File: gum/arch-arm/gumarmreg.c [hunk 13]
```diff
@@ -14,25 +14,47 @@ gum_arm_reg_describe (arm_reg reg,
...
+    ri->index = ri->meta - GUM_ARM_MREG_S0;
+  }
+  else if (reg >= ARM_REG_D0 && reg <= ARM_REG_D31)
+  {
+    ri->meta = GUM_ARM_MREG_D0 + (reg - ARM_REG_D0);
+    ri->width = 64;
+
```
Comment (oleavr) [inline gum/arch-arm/gumarmreg.c:52]:
```suggestion
```

File: gum/arch-arm/gumthumbwriter.c [hunk 14]
```diff
@@ -788,6 +788,50 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
...
+    code |= (dst_reg_info.index >> 1) << 12;
+    code |= (dst_reg_info.index & 1) << 22;
+  }
+  else 
+  {
+    code |= (dst_reg_info.index) << 12;
+    code |= 1 << 8;
+  }
+  
+  gum_thumb_writer_put_instruction_wide(self, code >> 16, code & 0xffff);
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:830]:
```suggestion
  gum_thumb_writer_put_instruction_wide (self, code >> 16, code & 0xffff);
```

File: gum/arch-arm/gumthumbwriter.c [hunk 15]
```diff
@@ -788,6 +788,50 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
...
+  GumArmRegInfo dst_reg_info;
+
+  // cccc cccc cycc ssss dddd cccx oooo oooo
+  // c: opt code fixed
+  // y: dest reg index, bit 1 
+  // s: source reg index
+  // d: dest reg index, bit 2-5
+  // x: 0 f32 load  1 f64 load
+
+  uint32_t code = 0xed900a00;
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:807]:
```suggestion
  guint32 code = 0xed900a00;
```

File: gum/arch-arm/gumthumbwriter.c [hunk 16]
```diff
@@ -788,6 +788,50 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
...
+  gboolean is_float = TRUE;
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:808]: These two variables should be moved up right after the other two variable declarations.

File: tests/core/arch-arm/thumbrelocator.c [hunk 17]
```diff
@@ -84,6 +86,89 @@ TESTCASE (handle_extended_instructions)
   g_assert_cmpint (memcmp (fixture->output, input, sizeof (input)), ==, 0);
 }
 
+TESTCASE (vldrpc_t1_should_be_rewritten)
```
Comment (oleavr) [inline tests/core/arch-arm/thumbrelocator.c:89]: Should be placed immediately after `ldrpc_t2_should_be_rewritten`.

**Comment by bigboysun** [comment]
check please

**Comment by bigboysun** [comment]
In the process of dealing with this problem, it is found that the lower version of "capstone" cannot deal with instructions such as "Vsub. F64". It is time to synchronize capstone.

ee31 0b40   vsub.f64  d0, d1, d0
eeb4 0bc1   vcmpe.f64 d0, d1

File: gum/arch-arm/gumthumbrelocator.c [hunk 18]
```diff
@@ -583,6 +588,31 @@ gum_thumb_relocator_rewrite_ldr (GumThumbRelocator * self,
   return TRUE;
 }
 
+static gboolean
+gum_thumb_relocator_rewrite_vldr (GumThumbRelocator * self,
+                                 GumCodeGenCtx * ctx)
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.c:593]:
```suggestion
                                  GumCodeGenCtx * ctx)
```

File: gum/arch-arm/gumthumbrelocator.c [hunk 19]
```diff
@@ -583,6 +588,31 @@ gum_thumb_relocator_rewrite_ldr (GumThumbRelocator * self,
...
+  if (src->type != ARM_OP_MEM || src->mem.base != ARM_REG_PC)
+    return FALSE;
+
+  absolute_pc = ctx->pc & ~((GumAddress) (4 - 1));
+  absolute_pc += src->mem.disp;
+
+  gum_thumb_writer_put_push_regs(ctx->output, 1, ARM_REG_R0);
+
+  gum_thumb_writer_put_ldr_reg_address (ctx->output, ARM_REG_R0, absolute_pc);
+  gum_thumb_writer_put_vldr_reg_reg_offset (ctx->output, dst->reg, ARM_REG_R0,
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.c:608]:
```suggestion
  gum_thumb_writer_put_vldr_reg_reg_offset (ctx->output, dst->reg, ARM_REG_R0,
```

File: gum/arch-arm/gumthumbrelocator.c [hunk 20]
```diff
@@ -583,6 +588,31 @@ gum_thumb_relocator_rewrite_ldr (GumThumbRelocator * self,
...
+  const cs_arm_op * src = &ctx->detail->operands[1];
+  GumAddress absolute_pc;
+
+  if (src->type != ARM_OP_MEM || src->mem.base != ARM_REG_PC)
+    return FALSE;
+
+  absolute_pc = ctx->pc & ~((GumAddress) (4 - 1));
+  absolute_pc += src->mem.disp;
+
+  gum_thumb_writer_put_push_regs(ctx->output, 1, ARM_REG_R0);
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.c:605]:
```suggestion
  gum_thumb_writer_put_push_regs (ctx->output, 1, ARM_REG_R0);
```

File: gum/arch-arm/gumthumbwriter.c [hunk 21]
```diff
@@ -788,6 +788,55 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
       GUM_THUMB_MEMORY_LOAD, dst_reg, src_reg, src_offset);
 }
 
+gboolean
+gum_thumb_writer_put_vldr_reg_reg_offset (GumThumbWriter * self,
+                                         arm_reg dst_reg,
+                                         arm_reg src_reg,
+                                         gsize src_offset)
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:795]:
```suggestion
                                          arm_reg dst_reg,
                                          arm_reg src_reg,
                                          gsize src_offset)
```

File: gum/arch-arm/gumthumbwriter.c [hunk 22]
```diff
@@ -788,6 +788,55 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
...
+{
+  GumArmRegInfo src_reg_info;
+  GumArmRegInfo dst_reg_info;
+
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:799]:
```suggestion
```

File: gum/arch-arm/gumthumbwriter.c [hunk 23]
```diff
@@ -788,6 +788,55 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
...
+  gum_arm_reg_describe (src_reg, &src_reg_info);
+  gum_arm_reg_describe (dst_reg, &dst_reg_info);
+
+  code |= (src_offset >> 2) & 0xff;
+  code |= (src_reg_info.index << 16);
+
+  if (dst_reg_info.meta >= GUM_ARM_MREG_D0 
+      && dst_reg_info.meta <= GUM_ARM_MREG_D31)
+    is_float = FALSE;
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:822]:
```suggestion
  is_float = (dst_reg_info.meta >= GUM_ARM_MREG_S0 &&
      dst_reg_info.meta <= GUM_ARM_MREG_S31);
```

File: gum/arch-arm/gumthumbwriter.c [hunk 24]
```diff
@@ -788,6 +788,55 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
...
+gum_thumb_writer_put_vldr_reg_reg_offset (GumThumbWriter * self,
+                                         arm_reg dst_reg,
+                                         arm_reg src_reg,
+                                         gsize src_offset)
+{
+  GumArmRegInfo src_reg_info;
+  GumArmRegInfo dst_reg_info;
+
+  guint32 code = 0xed900a00;
+  gboolean is_float = TRUE;
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:801]:
```suggestion
  gboolean is_float;
```

File: gum/arch-arm/gumthumbwriter.c [hunk 25]
```diff
@@ -788,6 +788,55 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
...
+gboolean
+gum_thumb_writer_put_vldr_reg_reg_offset (GumThumbWriter * self,
+                                         arm_reg dst_reg,
+                                         arm_reg src_reg,
+                                         gsize src_offset)
+{
+  GumArmRegInfo src_reg_info;
+  GumArmRegInfo dst_reg_info;
+
+  guint32 code = 0xed900a00;
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:800]:
```suggestion
  guint32 insn = 0xed900a00;
```

(For consistency with how we've named it in other functions.)

File: gum/arch-arm/gumthumbwriter.c [hunk 26]
```diff
@@ -788,6 +788,55 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
...
+  /*
+   * vldr instruction format
+   * cccc cccc cycc ssss dddd cccx oooo oooo
+   * c: opt code fixed
+   * y: dest reg index, bit 1 
+   * s: source reg index
+   * d: dest reg index, bit 2-5
+   * x: 0 f32 load  1 f64 load
+   * o: offset
+   */
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:812]: Reading the code now, I feel like this comment is duplicating what the code is already explaining, so we should probably remove it.

File: gum/arch-arm/gumthumbwriter.c [hunk 27]
```diff
@@ -788,6 +788,55 @@ gum_thumb_writer_put_ldr_reg_reg_offset (GumThumbWriter * self,
...
+
+  gum_arm_reg_describe (src_reg, &src_reg_info);
+  gum_arm_reg_describe (dst_reg, &dst_reg_info);
+
+  code |= (src_offset >> 2) & 0xff;
+  code |= (src_reg_info.index << 16);
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.c:818]:
```suggestion
  code |= src_reg_info.index << 16;
```

**Comment by oleavr** [comment]
> In the process of dealing with this problem, it is found that the lower version of "capstone" cannot deal with instructions such as "Vsub. F64". It is time to synchronize capstone.

This was already done recently. Are you sure your `capstone` submodule has been updated locally?

File: tests/core/arch-arm/thumbrelocator.c [hunk 28]
```diff
@@ -154,6 +156,82 @@ TESTCASE (ldrpc_t2_should_be_rewritten)
       sizeof (expected_output)), ==, 0);
 }
 
+TESTCASE (vldrpc_t1_should_be_rewritten)
+{
+  const guint16 input[] = {
+    GUINT16_TO_LE (0xeddf),
```
Comment (oleavr) [inline tests/core/arch-arm/thumbrelocator.c:162]:
```suggestion
    GUINT16_TO_LE (0xeddf),
```

File: tests/core/arch-arm/thumbrelocator.c [hunk 29]
```diff
@@ -154,6 +156,82 @@ TESTCASE (ldrpc_t2_should_be_rewritten)
...
+  gum_thumb_writer_flush (&fixture->tw);
+  
+  g_assert_cmpint (memcmp (fixture->output, expected_output,
+      sizeof (expected_output)), ==, 0);
+}
+
+TESTCASE (vldrpc_t2_should_be_rewritten)
+{
+  const guint16 input[] = {
+    GUINT16_TO_LE (0xed9f),
```
Comment (oleavr) [inline tests/core/arch-arm/thumbrelocator.c:200]:
```suggestion
    GUINT16_TO_LE (0xed9f),
```

**Pull request merged**